File: source/ts/tests/unit/defaultLocation.test.ts
import { describe, it, expect } from 'vitest';
import {
  TRADITIONAL_REFERENCE,
  MODERN_REFERENCE,
  IST_OFFSET_MINUTES,
  IST_TIMEZONE,
  referenceLocation,
  resolveLocation,
} from '../../src/core/defaultLocation';
import { PanchangError } from '../../src/types/errors';

const PUNE = { latitude: 18.5204, longitude: 73.8567 };

describe('reference points', () => {
  it('TRADITIONAL_REFERENCE is the classical madhya rekha', () => {
    expect(TRADITIONAL_REFERENCE.latitude).toBe(23.1765);
    expect(TRADITIONAL_REFERENCE.longitude).toBe(75.7885);
  });

  it('MODERN_REFERENCE is the 1955 Calendar Reform Committee point', () => {
    expect(MODERN_REFERENCE.latitude).toBe(23.1833);
    expect(MODERN_REFERENCE.longitude).toBe(82.5);
  });

  it('only the Central Station has IST as its own local mean time', () => {
    expect(MODERN_REFERENCE.longitude * 4).toBe(IST_OFFSET_MINUTES);
    expect(TRADITIONAL_REFERENCE.longitude * 4).toBeLessThan(IST_OFFSET_MINUTES);
    expect(IST_OFFSET_MINUTES - TRADITIONAL_REFERENCE.longitude * 4).toBeCloseTo(26.8, 1);
    expect(IST_TIMEZONE).toBe('Asia/Kolkata');
  });

  it('are frozen, so one caller cannot move them for every other', () => {
    expect(Object.isFrozen(TRADITIONAL_REFERENCE)).toBe(true);
    expect(Object.isFrozen(MODERN_REFERENCE)).toBe(true);
  });
});

describe('referenceLocation', () => {
  it('maps the two selectable modes to their points', () => {
    expect(referenceLocation('traditional')).toBe(TRADITIONAL_REFERENCE);
    expect(referenceLocation('modern')).toBe(MODERN_REFERENCE);
  });

  it('rejects practical and anything unknown', () => {
    for (const bad of ['practical', 'nonsense', undefined]) {
      expect(() => referenceLocation(bad as never)).toThrow(PanchangError);
    }
  });
});

describe('resolveLocation', () => {
  it('defaults to the traditional reference when no location is given', () => {
    for (const absent of [undefined, null] as const) {
      const r = resolveLocation(absent);
      expect(r.reference).toBe('traditional');
      expect(r.location).toBe(TRADITIONAL_REFERENCE);
    }
  });

  it('honours an explicit modern reference', () => {
    const r = resolveLocation(undefined, 'modern');
    expect(r.reference).toBe('modern');
    expect(r.location).toBe(MODERN_REFERENCE);
  });

  it('reports a supplied location as practical, whatever mode was asked for', () => {
    for (const mode of ['traditional', 'modern'] as const) {
      const r = resolveLocation(PUNE, mode);
      expect(r.reference).toBe('practical');
      expect(r.location).toBe(PUNE);
    }
  });

  it('treats 0,0 as a location, not as missing', () => {
    const r = resolveLocation({ latitude: 0, longitude: 0 });
    expect(r.reference).toBe('practical');
    expect(r.location.latitude).toBe(0);
  });

  it('rejects a half-filled location instead of completing it from a reference point', () => {
    expect(() => resolveLocation({ latitude: 18.52, longitude: Number.NaN }))
      .toThrow(PanchangError);
    try {
      resolveLocation({ latitude: Number.NaN, longitude: 73.85 });
    } catch (e) {
      expect((e as PanchangError).code).toBe('INVALID_LATITUDE');
    }
  });
});

File: source/go/internal/core/defaultlocation_test.go
package core

import (
	"testing"

	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

func TestReferencePoints(t *testing.T) {
	if u := TraditionalReference(); u.Latitude != 23.1765 || u.Longitude != 75.7885 {
		t.Errorf("TraditionalReference() = %+v, want {23.1765 75.7885 0}", u)
	}
	s := ModernReference()
	if s.Latitude != 23.1833 || s.Longitude != 82.5 {
		t.Errorf("ModernReference() = %+v, want {23.1833 82.5 0}", s)
	}
	if got := s.Longitude * 4; got != ISTOffsetMinutes {
		t.Errorf("central station longitude*4 = %v, want %d", got, ISTOffsetMinutes)
	}
	if got := TraditionalReference().Longitude * 4; got >= ISTOffsetMinutes {
		t.Errorf("Ujjain LMT = %v, expected it west of IST's %d", got, ISTOffsetMinutes)
	}
}

func TestReferencePointsAreCopiedNotShared(t *testing.T) {
	u := TraditionalReference()
	u.Latitude = 0
	if TraditionalReference().Latitude != 23.1765 {
		t.Error("mutating a returned point changed the next caller's copy")
	}
}

func TestReferenceLocation(t *testing.T) {
	got, err := ReferenceLocation(ReferenceTraditional)
	if err != nil || got != TraditionalReference() {
		t.Errorf("traditional = %+v, %v; want the traditional point", got, err)
	}
	got, err = ReferenceLocation(ReferenceModern)
	if err != nil || got != ModernReference() {
		t.Errorf("modern = %+v, %v; want the Central Station", got, err)
	}
	if _, err := ReferenceLocation(ReferencePractical); err == nil {
		t.Error("ReferenceLocation(practical) must error; it is a reported frame, not a selectable one")
	}
	if _, err := ReferenceLocation(Reference("nonsense")); err == nil {
		t.Error("an unknown mode must error")
	}
}

func TestResolveLocation(t *testing.T) {
	got, ref, err := ResolveLocation(nil, ReferenceTraditional)
	if err != nil || ref != ReferenceTraditional || got != TraditionalReference() {
		t.Errorf("nil traditional: %+v %q %v; want the traditional point", got, ref, err)
	}

	got, ref, err = ResolveLocation(nil, ReferenceModern)
	if err != nil || ref != ReferenceModern || got != ModernReference() {
		t.Errorf("nil modern: %+v %q %v; want the station and modern", got, ref, err)
	}

	pune := types.GeoLocation{Latitude: 18.5204, Longitude: 73.8567}
	got, ref, err = ResolveLocation(&pune, ReferenceTraditional)
	if err != nil || ref != ReferencePractical || got != pune {
		t.Errorf("supplied: %+v %q %v; want Pune and practical", got, ref, err)
	}

	zero := types.GeoLocation{}
	got, ref, err = ResolveLocation(&zero, ReferenceTraditional)
	if err != nil || ref != ReferencePractical || got != zero {
		t.Errorf("zero value: %+v %q %v; want Null Island and practical", got, ref, err)
	}

	bad := types.GeoLocation{Latitude: 91}
	if _, _, err := ResolveLocation(&bad, ReferenceTraditional); err == nil {
		t.Error("an out-of-range latitude must error rather than fall back to a reference point")
	}
}
